executor/adapter.go

```diff
@@ -1480,15 +1480,37 @@ func resetCTEStorageMap(se sessionctx.Context) error {
 // LogSlowQuery is used to print the slow query in the log files.
 func (a *ExecStmt) LogSlowQuery(txnTS uint64, succ bool, hasMoreResults bool) {
 	sessVars := a.Ctx.GetSessionVars()
+	// Disable recording the internal queries.
+	if sessVars.ConnectionID == 0 {
+		return
+	}
 	stmtCtx := sessVars.StmtCtx
 	level := log.GetLevel()
 	cfg := config.GetGlobalConfig()
 	costTime := time.Since(sessVars.StartTime) + sessVars.DurationParse
 	threshold := time.Duration(atomic.LoadUint64(&cfg.Instance.SlowThreshold)) * time.Millisecond
+	expensiveRatio := atomic.LoadUint64(&cfg.Instance.ExpensiveScanRatio)
 	enable := cfg.Instance.EnableSlowLog.Load()
+
+	// Check if it's an expensive mvcc scan. It's regarded as expensive if
+	// 1. The rocksdb_delete_skipped_count > processed_keys * expensive_ratio
+	// 2. The processed_keys is 0 but the duration is larger than slow threshold.
+	execDetail := stmtCtx.GetExecDetails()
+	isExpensiveScan := false
+	if execDetail.ScanDetail != nil && sessVars.ConnectionID > 0 {
+		processedKeys := execDetail.ScanDetail.ProcessedKeys
+		if processedKeys <= 0 {
+			processedKeys = 1
+		}
+		currentScanRatio := execDetail.ScanDetail.RocksdbKeySkippedCount / uint64(processedKeys)
```

Comment:
In the case of processed keys = 1 and total scanned keys are 100, it will also be logged, I think we will receive a log of annoying logs. How about also consider total scan keys count threshold?

Reply:
I agree.  I think a scan keys count threshold is necessary to avoid flooding logs.